### DISCUSSION TITLE
Missing types while migrating from 3.x to 4.x

### DISCUSSION THREAD

**Discussion by nulltoken** [panva/jose#293] [Q&A] [answered]
### What happened?

While migrating to the latest 4.x, we discovered some TypeScript type definitions are no longer exported.

We using your library in a context where we're working with many trusted authorization servers.

Mainly, we've created a higher order keyRetriever typed `GetKeyFunction<JWSHeaderParameters, FlattenedJWSInput>` that 
decodes the jwt payload and depdending on the 'iss' claim returns the proper remote JWKSFetcher.

Previously we were relying on:

```
import { FlattenedJWSInput, GetKeyFunction, JWSHeaderParameters, JWTPayload } from 'jose/types';
import { decode } from 'jose/util/base64url';
```

And it seems `base64url.decode` and `GetKeyFunction` are no longer accessible through the main 'jose' export.

Would you be open to a PullRequest exporting those?

### Version

4.1.0

### Runtime

Node.js

### Runtime Details

v14.18.0

### Code to reproduce

```js
Not really applicable.
```

### 

- [X] I have searched the issues tracker and discussions for similar topics and couldn't find anything related.
- [X] I agree to follow this project's [Code of Conduct](https://github.com/panva/jose/blob/main/CODE_OF_CONDUCT.md)

**Comment by panva** [comment, marked as answer]
https://github.com/panva/jose/releases/tag/v4.0.0

### Migration Guide

Migrating from v3.x to v4.x is very straight forward. 

1. All `import` statements, `require()`, or `import()` invocations should be changed to use just `'jose'` as the target.
2. There are no more default exports. Everything was converted to a named export.
3. Refer to the [documentation](https://github.com/panva/jose/tree/v4.x/docs) if you're not sure how a given module should be imported / required.
4. any custom resolvers targeting `jose` dist files for jest, typescript, or other tooling should be removed

```js
// before (v3.x)
import { jwtVerify } from 'jose/jwt/verify'
import { SignJWT } from 'jose/jwt/sign'
import * as errors from 'jose/util/errors'

// after (v4.x)
import { jwtVerify, SignJWT, errors } from 'jose'
```

---

Furthermore, all types are exported from the actual main `jose` export that was not present in v3.x at all. The exception is `GetKeyFunction` which instead is exported with the actual types for individual functions > `CompactDecryptGetKey`, `FlattenedDecryptGetKey`, `GeneralDecryptGetKey`, `CompactVerifyGetKey`, `FlattenedVerifyGetKey`, `GeneralVerifyGetKey`, `JWTVerifyGetKey`, `JWTDecryptGetKey`.

https://github.com/panva/jose/blob/v4.1.0/src/index.ts

base64url is also exported.

**Reply by gproanomsft** [reply]
how does this work with jose-node-cjs-runtime?  i'm no longer able to import JWSSignatureVerificationFailed, JWTClaimValidationFailed, JWTExpired

do I need to use the jose package instead?

**Reply by panva** [reply]
Specific runtimes play absolutely no role in the types or exports.

**Reply by gproanomsft** [reply]
i get an error when I try to import from jose while using the jose-node-cjs-runtime package

**Reply by panva** [reply]
If using a different node module name confuses you, you might as well just install and use the `jose` identifier.

**Comment by nulltoken** [comment]
@panva That last paragraph was the missing link from https://github.com/panva/jose/releases/tag/v4.0.0

Awesome feedback! Thanks a lot.
